extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts

```diff
@@ -0,0 +1,500 @@
+
+import * as azdata from 'azdata';
+import * as nls from 'vscode-nls';
+import { ClusterController, IEndPoint, IControllerError } from '../controller/clusterController';
+import { ControllerTreeDataProvider } from '../tree/controllerTreeDataProvider';
+import { TreeNode } from '../tree/treeNode';
+
+const localize = nls.loadMessageBundle();
+
+export class AddControllerDialogModel {
+	constructor(
+		public treeDataProvider: ControllerTreeDataProvider,
+		public node?: TreeNode,
+		public prefilledUrl?: string,
+		public prefilledUsername?: string,
+		public prefilledPassword?: string,
+		public prefilledRememberPassword?: boolean
+	) {
+		this.prefilledUrl = prefilledUrl || (node && node['url']);
+		this.prefilledUsername = prefilledUsername || (node && node['username']);
+		this.prefilledPassword = prefilledPassword || (node && node['password']);
+		this.prefilledRememberPassword = prefilledRememberPassword || (node && node['rememberPassword']);
+	}
+
+	public async onComplete(url: string, username: string, password: string, rememberPassword: boolean): Promise<void> {
+		let clusterController = new ClusterController();
+		let response = await clusterController.getEndPoints(url, username, password, true);
+		if (response && response.request) {
+			let masterInstance: IEndPoint = undefined;
+			if (response.endPoints) {
+				masterInstance = response.endPoints.find(e => e.name && e.name === 'sql-server-master');
+			}
+			this.treeDataProvider.addController(response.request.url, response.request.username,
+				response.request.password, rememberPassword, masterInstance);
+			await this.treeDataProvider.saveControllers();
+		}
+	}
+
+	public async onError(error: IControllerError): Promise<void> {
+		// implement
+	}
+
+	public async onCancel(): Promise<void> {
+		if (this.node) {
+			this.node.refresh();
+		}
+	}
+}
+
+export class AddControllerDialog {
+
+	private readonly DialogTitle: string = localize('textAddNewController', 'Add New Controller');
+	private readonly SignInButtonText: string = localize('textAdd', 'Add');
+	private readonly CancelButtonText: string = localize('textCancel', 'Cancel');
+
+	private dialog: azdata.window.Dialog;
+	private uiModelBuilder: azdata.ModelBuilder;
+
+	private urlInputBox: azdata.InputBoxComponent;
+	private usernameInputBox: azdata.InputBoxComponent;
+	private passwordInputBox: azdata.InputBoxComponent;
+	private rememberPwCheckBox: azdata.CheckBoxComponent;
+	private errorTextContainer: azdata.FlexContainer;
+
+	constructor(private model: AddControllerDialogModel) {
+	}
+
+	public showDialog(): void {
+		this.createDialog();
+		this.toggleSignInButton();
+		this.toggleErrorText();
+		azdata.window.openDialog(this.dialog);
+	}
+
+	private createDialog(): void {
+		this.dialog = azdata.window.createModelViewDialog(this.DialogTitle);
+		this.dialog.registerContent(async view => {
+			this.uiModelBuilder = view.modelBuilder;
+
+			let titledContainer = new TitledContainer(this.uiModelBuilder);
```

Comment:
I'm confused why we're not using a standard Form layout here. If we have limitations let's raise with larger team, but it seems like we have a lot of custom CSS here for what should be a standard form. We've had to create a special TitledContainer class plus lots of flex magic - is this really necessary?

Reply:
I think we covered all the scenarios offline - let's go with FormContainer

Reply:
fixed.